OrbitService/ProcessList.cpp

```diff
@@ -41,66 +33,43 @@ ErrorMessageOr<void> ProcessList::Refresh() {
     uint32_t pid;
     if (!absl::SimpleAtoi(folder_name, &pid)) continue;
 
-    auto iter = processes_map_.find(pid);
-    if (iter != processes_map_.end()) {
-      ProcessInfo& process(*(iter->second));
-      process.set_cpu_usage(cpu_usage_map[process.pid()]);
-      updated_processes.push_back(process);
-      continue;
-    }
+    const auto iter = processes_.find(pid);
+    if (iter != processes_.end()) {
+      auto process = processes_.extract(iter);
 
-    const std::filesystem::path name_file_path = path / "comm";
-    auto name_file_result = utils::ReadFileToString(name_file_path);
-    if (!name_file_result) {
-      ERROR("Failed to read %s: %s", name_file_path.string(), name_file_result.error().message());
-      continue;
-    }
-    std::string name = std::move(name_file_result.value());
-    // Remove new line character.
-    absl::StripTrailingAsciiWhitespace(&name);
-    if (name.empty()) continue;
-
-    ProcessInfo process;
-    process.set_pid(pid);
-    process.set_name(name);
-    process.set_cpu_usage(cpu_usage_map[pid]);
-
-    // "The command-line arguments appear [...] as a set of strings
-    // separated by null bytes ('\0')".
-    const std::filesystem::path cmdline_file_path = directory_entry.path() / "cmdline";
-    auto cmdline_file_result = utils::ReadFileToString(cmdline_file_path);
-    if (!cmdline_file_result) {
-      ERROR("Failed to read %s: %s", cmdline_file_path.string(),
-            name_file_result.error().message());
+      const auto total_cpu_time = utils::GetCumulativeTotalCpuTime();
+      const auto cpu_time = utils::GetCumulativeCpuTimeFromProcess(process.value().pid());
+      if (cpu_time && total_cpu_time) {
+        process.value().UpdateCpuUsage(cpu_time.value(), total_cpu_time.value());
+      } else {
+        // We don't fail in this case. This could be a permission problem which might occur when not
+        // running as root.
+        ERROR("Could not update the CPU usage of process %d", process.value().pid());
+      }
+
+      updated_processes.insert(std::move(process));
```

Comment:
Should this maybe only be done when we succeed in updating the cpu usage? Just an idea. Feel free to dismiss

Reply:
Good question. I thought about these kind of things and decided that I don't want to change behaviour in this PR. Not reading CPU usage is something which usually happens only when the service is not running as root which is currently out of scope. We should probably think about how to handle that in the future, things like differentiating between processes that can be profiled and these that can't.

Reply:
Sounds good